substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVLineRecordBuilder.java

```diff
@@ -0,0 +1,175 @@
+ * Copyright (c) 2020, 2020, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.objectfile.pecoff.cv;
+
+import org.graalvm.compiler.debug.DebugContext;
+import com.oracle.objectfile.debugentry.FileEntry;
+import com.oracle.objectfile.debugentry.PrimaryEntry;
+import com.oracle.objectfile.debugentry.Range;
+
+/*
+ * In CV4, the line table consists of a series of file headers followed by line number entries.
+ * to handle this, first we decide if we want to merge this with the previous range (only if
+ * same file and start of this range is end of previous range) if we are emitting a new range to
+ * the same file, write the range, save it as the previous range and go on If this is a
+ * different file, then update the length of the previous file header, write the new file header
+ * and write the new range At the very end, make sure we update the last file header.
+ *
+ * In addition, optionally ignore Ranges that point into Graal innards, just adding them to the
+ * current enclosing range.
+ */
+
+public class CVLineRecordBuilder {
+
+    private CVDebugInfo cvDebugInfo;
+    private DebugContext debugContext;
+    private CVLineRecord lineRecord;
+    private PrimaryEntry primaryEntry;
+
+    CVLineRecordBuilder(DebugContext theDebugContext, CVDebugInfo cvDebugInfo) {
+        this.debugContext = theDebugContext;
+        this.cvDebugInfo = cvDebugInfo;
+    }
+
+    public void debug(String format, Object... args) {
+        cvDebugInfo.getCVSymbolSection().verboseLog(debugContext, format, args);
+    }
+
+    /**
+     * Build line number records for a function.
+     *
+     * @param entry function to build line number table for
+     * @return CVLineRecord containing any entries generated, or null if no entries generated
+     */
+    CVLineRecord build(PrimaryEntry entry, String methodName) {
```

Comment:
I'd rather not; I don't think the code would be much clearer (but it would be less efficient - I'd like to rewrite it more of a stream handler), and there's a chance that the Builder might in future need to know the previous state when I reimplement some of the code that I've removed.